rts/Rendering/Textures/Bitmap.cpp

```diff
@@ -1922,14 +1922,29 @@ void CBitmap::CopySubImage(const CBitmap& src, int xpos, int ypos)
 
 	const uint8_t* srcMem = src.GetRawMem();
 	      uint8_t* dstMem =     GetRawMem();
-
 	const auto dts = GetDataTypeSize();
-	for (int y = 0; y < src.ysize; ++y) {
-		const int pixelDst = (((ypos + y) *     xsize) + xpos) * channels * dts;
-		const int pixelSrc = ((        y  * src.xsize) +    0) * channels * dts;
 
-		// copy the whole line
-		std::copy(&srcMem[pixelSrc], &srcMem[pixelSrc] + src.xsize * channels * dts, &dstMem[pixelDst]);
+	if (src.channels == channels) {
+		for (int y = 0; y < src.ysize; ++y) {
+			const int pixelDst = (((ypos + y) *     xsize) + xpos) * channels * dts;
+			const int pixelSrc = ((        y  * src.xsize) +    0) * channels * dts;
+
+			// copy the whole line
+			std::copy(&srcMem[pixelSrc], &srcMem[pixelSrc] + src.xsize * channels * dts, &dstMem[pixelDst]);
+		}
+	} else if (src.channels == 1 && channels == 4 && dts == 1 && src.GetDataTypeSize() == 1) {
+		// set rgb to 255 and alpha to src tex value
+		constexpr uint32_t baseColor = ((255<<16)|(255<<8)|(255));
```

Comment:
Won't this depend on the color format? e.g. with RGBA vs ARGB you'd have <<24, <<16 and <<8 instead

Reply:
right, will check for that as well

Reply:
* This should become a `BitmapAction`.
* I don't think we need to support anything but `RGBA` in `CBitmap`, if it's ARGB or other abracadabra then it's up for a caller to normalize it, or do another Bitmap/BitmapAction routine to shuffle.

Reply:
@lhog we can introduce more actions for this (like CopyConvertSubImage), and have CopySubImage work only for same format bitmaps. But using several actions when you can use one I'm not so sure it's best, in many cases if you can do it in just one action it's going to be more direct and likely more efficient.

For example when copying RGBA to BGRA or ARGB, you can shuffle first and then CopySubImage, but with the right algo it can be done in one go.

Likewise, in this case we're doing Grayscale to RGBA/BGRA where something similar happens, the Grayscale can be converted first to color, but here it's converted and the result placed directly at destination.

Let me know your final decision and I'll resolve this.

Reply:
I see what's the problem here, CBitmap doesn't have a pixel format field :P, so it knows nothing of RGBA vs ARGB (or RGBA vs BGRA, but that's not really an issue here).

Also, the CBitmap class has other actions expecting alpha in the same place.

It does support 4 and 1 channels, so imo it should be fine to allow CBitmap(4channels).CopySubimage(cbitmap(1channel)) like I'm doing, at least while a more comprehensive solution for this is developed (I think an action system supporting two bitmaps with specific pixel formats could be designed so we can select the right action when two bitmaps are combined with different pixel formats).

If you really don't want this in there now, I propose something like cbitmap.ImportSubImage(cbitmap) that would work like CopySubImage but importing different formats. It's basically the same thing in the end though.

Otherwise, if you really want to have an action+copysubimage for this, I still think it's better to avoid that since we would do more processing in a pretty critical place mostly for esthetics. But ofc it can be done. In that case I think freetype actually has [something](https://freetype.org/freetype2/docs/reference/ft2-bitmap_handling.html#ft_bitmap_convert) to convert the bitmap to different channel number, so I can probably directly use that and it's likely that method is well optimized too.